api/internal/git/repospec.go

```diff
@@ -252,22 +268,32 @@ func parseHostSpec(n string) (string, string) {
 		}
 	}
 
-	return normalizeGitHostSpec(host), n
+	host, err := normalizeGitHostSpec(host)
+	return host, n, err
 }
 
-func normalizeGitHostSpec(host string) string {
+var userRegex = regexp.MustCompile(`^(?:ssh://|git::)?([a-zA-Z][a-zA-Z0-9-]*@)`)
+
+func normalizeGitHostSpec(host string) (string, error) {
 	s := strings.ToLower(host)
+	m := userRegex.FindStringSubmatch(host)
 	if strings.Contains(s, "github.com") {
-		if strings.Contains(s, "git@") || strings.Contains(s, "ssh:") {
+		switch {
+		case len(m) > 0:
+			if strings.HasPrefix(s, "git::") && m[1] != "git@" {
+				return "", fmt.Errorf("git protocol on github.com only allows git@ user")
+			}
+			host = m[1] + "github.com:"
+		case strings.Contains(s, "ssh:"):
 			host = "git@github.com:"
-		} else {
+		default:
 			host = "https://github.com/"
 		}
 	}
```

Comment:
Suggest not use "switch/case".

Reason: "switch/case" is read as the conditions are comparable or in parallel. They should be considered as one or the other (this is also why ["fallthrough" is arguable](https://stackoverflow.com/questions/188461/switch-statement-fall-through-should-it-be-allowed)). I don't think this is what your code is trying to deliver, so it's better to not use switch/case to make the code more readability. :)

Reply:
I'm not sure if I understand this statement. Go switch statements do not behave like that of other languages. In Go, a switch statement behaviors like a series of `if/else` statements, which this is replacing.

Are you saying that we should avoid `switch` in general because it behaves differently in Go from other languages and can be confusing?

Reply:
Sorry for the confusion. I'm reviewing the PR from the readability perspective (whether the code itself can be self-explained and easy to maintain). That's said, the switch/case here can be used as if/else and the logic is 100% correct, but it seems not very straightforward to follow.

Reply:
I agree with @yuwenma's comment from readability perspective here.

Reply:
Would it be easier to read and maintain the code if we do not use regex (or regex group, but go does not have good support for regex group).
For example, 
```go
func normalizeGitHostSpec(host string) (string, error) {
	s := strings.ToLower(host)
	if !strings.Contains(s, "github.com") {
		// shall we check the "ssh::" as well?
		return strings.TrimPrefix(s, "git::"), nil
	}
        //either switch/case or if condition, defer to you. 
	switch true {
	case s == githubSSH:
	  return githubSSH, nil
	case strings.HasPrefix(s, "git::"):
	  s = strings.TrimPrefix(s,  "git::")
	  user, _ := strings.Split(s, "@")
	  if user != "git" {
	     return "", fmt.Errorf("git protocol on github.com only allows git@ user")
	  }
	  return githubSSH
	case strings.HasPrefix(s, "ssh://"):
	  user, _ := strings.Split(s, "ssh://")
          return user + "github.com:", nil
	case strings.HasPrefix(s, "ssh:"):
	  return githubSSH
	default:
          return githubHTTP
	}
``` 
The main concern I have is that the current code share git protocal path. Basically it treats `git::` and `ssh://` together, and handles `git::` with an additional constraints ( only support git@ user) inside the shared logic, and then handles `ssh:` and `git::` again in the following if conditions. If future features need to change the `git::` logic, it will affect the path of `ssh://` as well.

Reply:
Ok I've changed it to if/else!

I chosen to use regex here also because it can help with reducing user confusion, which is equally important as code readability. It requires that the format be `<protocol><identifier>@`. I can imagine the user adding `@` at the end to specify a ref, like `ssh://github.com/sig-kube/kustomize.git@master`, and getting a strange error about `github.com/sig-kube/kustomize.git@` being used as part of the username.

Regarding future changes, I think the current code is not too bad? It's only a few lines long and we can adapt as needed later.

Reply:
Ah, I remember why I wrote this as a switch now

```
Error: api/internal/git/repospec.go:300:3: ifElseChain: rewrite if-else to switch statement (gocritic)
		if len(m) > 0 {
```

Reply:
See the above switch/case readability suggestions. It could be error prone to maintain since `ssh::` and `git:` shares the condition check and handles the corner cases differently.

Reply:
Just to make myself clearer, I think your code is great and bug free. But since it adds a new path on top of the existing ssh logic, the code becomes more complicated than necessary. So I think some additional refactoring is needed. This helps both us (as maintainer) to better support your code, and for future contributors to easily understand and not mess it up.

Reply:
> The main concern I have is that the current code shares the git protocal logic paths. Basically it treats git:: and ssh:// together, and handles git:: with an additional constraints ( only support git@ user) inside the shared logic, and then handles ssh: and git:: again in the following if conditions.
> See the above switch/case readability suggestions. It could be error prone to maintain since ssh:: and git: shares the condition check and handles the corner cases differently.

I did try pretty hard to understand these statements, but I don't quite understand. I don't see the "then handles ssh: and git:: again in the following if conditions" part happening here. Are you saying to the `case` and `default` clause below the `case len(m) > 0`?

In my opinion, your suggested option has 5 if/else cases, has slightly more complexity in each case, and is thus less readable and more difficult to maintain. The option I currently have implemented has a rather simple regex, slightly simpler if/else cases, and only 3 _mutually exclusive_ cases.

Reply:
I apologize if I am being too dense here, but could you confirm that the changes made in this commit is functionally identical before and after?
https://github.com/kubernetes-sigs/kustomize/pull/4741/commits/b84d5971ff93538cddcbe2bb9e322ab89a1a5569

Reply:
"then handles ssh: and git:: again in the following if conditions" --> I mean the second `switch/case` (`ssh:`)  condition and the if statement after the `switch/case`  (ln291-293)

> In my opinion, your suggested option has 5 if/else cases, has slightly more complexity in each case, and is thus less readable and more difficult to maintain. 

Feel free to give other code structure solutions. Basically this ssh conditions are complicated, so I'm leaning towards listing each individual case rather than shorter lines of code. It's the similar debate about 
`return a & b & c & d & e` vs 
```
if !a:
   return false
if !b
   return false
if !c: 
   return false  
if !d:
   return false
if !e:
   return false
return true
```
the one-line code seems to be pretty, but it's always the latter that is easier to maintain and read.

Reply:
I wasn't intending to distract the PR review to this discussion that much. 
the original code is quite easy to read
```go
if strings.Contains(s, "git@") || strings.Contains(s, "ssh:") {
     host = "git@github.com:"
} else {
     host = "https://github.com/"
}
```
I saw that your request is to support one  more case, like `ssh://other-org@github.com`. If it's not convincing enough for you to change the code, can you make the minimum change like (psycho code)
```
if strings.Contains(s, "git@") || strings.Contains(s, "ssh:") {
    org := allowOtherOrg(s) 
    if org != "" {
        host = org  + "@github.com:
    } else {
        host = "git@github.com:"
    }
} else {
     host = "https://github.com/"
}
def allowOtherOrg(s string) string {
    // regex for the exact pattern your want to add "^ssh://[a-zA-Z][a-zA-Z0-9-]*@"
    // return the org
}
```

Reply:
Thank you for suggesting the readability improvements. I agree that it would be best to make sure the code is as simple and maintainable for the maintainers and future contributors.

I do believe though, the current code is a pretty simple option. Any changes to be made should both make the code simpler _and_ preserve the logic. I think the options suggested so far does one or the other, but not both.

The cases I would like to cover are:

1. GOOD - `org-12345@github.com`
2. GOOD - `ssh://org-12345@github.com`
3. GOOD - `git@github.com`
4. GOOD - `ssh://git@github.com`
5. GOOD - `git::git@github.com`
6. BAD - `git::org-12345@github.com`

Your newest suggestion would support cases 2-5, but would not support case 1. I think would also produce a strange result for case 6 and will return `https://github.com/`.

<br/>

In your other suggestion with 5 case statements, there are also a couple things I think are lacking from the maintainability and correctness perspective. I mentioned above but didn't go into detail but the following statement

```
	case strings.HasPrefix(s, "ssh://"):
	  user, _ := strings.Split(s, "@") // I changed this from "ssh://", assume this is what you intended
          return user + "github.com:", nil
```

would handle the good case of `ssh://org-12345@github.com`, producing `org-12345` as the user. However it would also handle the bad, but common case of `ssh://github.com/sig-kubernetes/kustomize@master`, which would produce user `github.com/sig-kubernetes/kustomize` and host `github.com/sig-kubernetes/kustomize@github.com`.

I personally think that additional branching is more cognitive load for the reader than a small regular expression like `^(?:ssh://|git::)?([a-zA-Z][a-zA-Z0-9-]*@`. This is actually why I went against @KnVerey's suggestion of using `\A` above. Using `\A` would be just as correct but I believe would reduce readability, just so ever slightly.

Regarding readability of the `switch` statement and whether it should be used, using `switch` in place of `if/else` statements is idiomatic Go. I acknowledge this may not be the case in Java/C++, but it is a common pattern in Golang. It is in [Effective Go](https://go.dev/doc/effective_go#switch). It is also commonly used within the Go standard library, like [here](https://github.com/golang/go/blob/9aa7107cb58417db4b19ea5f8773875a418270f3/src/database/sql/convert.go#L169) and [here](https://github.com/golang/go/blob/19309779ac5e2f5a2fd3cbb34421dafb2855ac21/src/internal/fmtsort/sort.go#L86). There are many other use cases. The Go standard library is widely considered a reference for Go style.

I think we both have the same goals and I appreciate your suggestions. I would also like to keep this codebase maintainable for future authors. I believe what is currently in this PR is good enough. If you or maybe @natasha41575 would be able to suggest an alternative that is logically equivalent and also cognitively simpler, I would be happy to make the changes.

Reply:
Made an update to reduce nested branching. I think it ends up being more complicated but maybe it's more readable.

Reply:
> Your newest suggestion would support cases 2-5, but would not support case 1. I think would also produce a strange result for case 6 and will return https://github.com/ instead of failing

Gotcha. My suggestion is from reading the previous and this PR's code intention. This is why I feel like the code logic is kind of twisted and need some readability improvements. 

I think your newest code does very good job on the variable names. They are meaningful and explains the code very well. @KnVerey also has some nice suggestions about comments and regex. 

Besides that, this PR looks good to me.